{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import json\n",
    "import datetime\n",
    "import numpy as np\n",
    "import skimage.draw\n",
    "\n",
    "import mrcnn\n",
    "from mrcnn.visualize import display_instances, display_top_masks\n",
    "from mrcnn.utils import extract_bboxes\n",
    "\n",
    "from mrcnn.utils import Dataset\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "from mrcnn.config import Config\n",
    "from mrcnn.model import MaskRCNN\n",
    "\n",
    "\n",
    "from mrcnn import model as modellib, utils\n",
    "from PIL import Image, ImageDraw\n",
    "import skimage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLASS_NAMES = [\"BG\", \"Kiste\", \"Packung\", \"Packung (oben)\"]\n",
    "\n",
    "# define the prediction configuration\n",
    "class PredictionConfig(Config):\n",
    "\t# define the name of the configuration\n",
    "\tNAME = \"inference_cfg\"\n",
    "\t# number of classes (background + Kiste + Packung + Packung (oben))\n",
    "\tNUM_CLASSES = len(CLASS_NAMES)\n",
    "\t# Set batch size to 1 since we'll be running inference on\n",
    "            # one image at a time. Batch size = GPU_COUNT * IMAGES_PER_GPU\n",
    "\tGPU_COUNT = 1\n",
    "\tIMAGES_PER_GPU = 1\n",
    "\t# BACKBONE = \"resnet50\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Configurations:\n",
      "BACKBONE                       resnet101\n",
      "BACKBONE_STRIDES               [4, 8, 16, 32, 64]\n",
      "BATCH_SIZE                     1\n",
      "BBOX_STD_DEV                   [0.1 0.1 0.2 0.2]\n",
      "COMPUTE_BACKBONE_SHAPE         None\n",
      "DETECTION_MAX_INSTANCES        35\n",
      "DETECTION_MIN_CONFIDENCE       0.7\n",
      "DETECTION_NMS_THRESHOLD        0.3\n",
      "FPN_CLASSIF_FC_LAYERS_SIZE     1024\n",
      "GPU_COUNT                      1\n",
      "GRADIENT_CLIP_NORM             5.0\n",
      "IMAGES_PER_GPU                 1\n",
      "IMAGE_CHANNEL_COUNT            3\n",
      "IMAGE_MAX_DIM                  1024\n",
      "IMAGE_META_SIZE                16\n",
      "IMAGE_MIN_DIM                  1024\n",
      "IMAGE_MIN_SCALE                0\n",
      "IMAGE_RESIZE_MODE              square\n",
      "IMAGE_SHAPE                    [1024 1024    3]\n",
      "LEARNING_MOMENTUM              0.9\n",
      "LEARNING_RATE                  0.001\n",
      "LOSS_WEIGHTS                   {'rpn_class_loss': 1.0, 'rpn_bbox_loss': 1.0, 'mrcnn_class_loss': 1.0, 'mrcnn_bbox_loss': 1.0, 'mrcnn_mask_loss': 1.0}\n",
      "MASK_POOL_SIZE                 14\n",
      "MASK_SHAPE                     [28, 28]\n",
      "MAX_GT_INSTANCES               100\n",
      "MEAN_PIXEL                     [123.7 116.8 103.9]\n",
      "MINI_MASK_SHAPE                (56, 56)\n",
      "NAME                           inference_cfg\n",
      "NUM_CLASSES                    4\n",
      "POOL_SIZE                      7\n",
      "POST_NMS_ROIS_INFERENCE        1000\n",
      "POST_NMS_ROIS_TRAINING         2000\n",
      "PRE_NMS_LIMIT                  6000\n",
      "ROI_POSITIVE_RATIO             0.33\n",
      "RPN_ANCHOR_RATIOS              [0.5, 1, 2]\n",
      "RPN_ANCHOR_SCALES              (32, 64, 128, 256, 512)\n",
      "RPN_ANCHOR_STRIDE              1\n",
      "RPN_BBOX_STD_DEV               [0.1 0.1 0.2 0.2]\n",
      "RPN_NMS_THRESHOLD              0.7\n",
      "RPN_TRAIN_ANCHORS_PER_IMAGE    256\n",
      "STEPS_PER_EPOCH                1000\n",
      "TOP_DOWN_PYRAMID_SIZE          256\n",
      "TRAIN_BN                       False\n",
      "TRAIN_ROIS_PER_IMAGE           200\n",
      "USE_MINI_MASK                  False\n",
      "USE_RPN_ROIS                   True\n",
      "VALIDATION_STEPS               10\n",
      "WEIGHT_DECAY                   0.0001\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# create config\n",
    "cfg = PredictionConfig()\n",
    "cfg.display() \n",
    "# define the model\n",
    "model = MaskRCNN(mode='inference', model_dir='logs', config=cfg)\n",
    "# load model weights\n",
    "model.load_weights('logs/kaggle/mask_rcnn_custom_cfg_coco_0003.h5', by_name=True, exclude=[\"mrcnn_class_logits\", \"mrcnn_bbox_fc\", \"mrcnn_mask\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing 1 images\n",
      "image                    shape: (480, 848, 3)         min:    0.00000  max:  255.00000  uint8\n",
      "molded_images            shape: (1, 1024, 1024, 3)    min: -123.70000  max:  151.10000  float64\n",
      "image_metas              shape: (1, 16)               min:    0.00000  max: 1024.00000  float64\n",
      "anchors                  shape: (1, 261888, 4)        min:   -0.35390  max:    1.29134  float32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-11 13:00:10.730440: W tensorflow/c/c_api.cc:305] Operation '{name:'mrcnn_detection_1/Reshape_1' id:16897 op device:{requested: '', assigned: ''} def:{{{node mrcnn_detection_1/Reshape_1}} = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _has_manual_control_dependencies=true](mrcnn_detection_1/packed, mrcnn_detection_1/Reshape_1/shape)}}' was changed by setting attribute after it was run by a session. This mutation will have no effect, and will trigger an error in the future. Either don't modify nodes after running them or create a new session.\n",
      "2024-07-11 13:00:10.893516: W tensorflow/c/c_api.cc:305] Operation '{name:'mrcnn_bbox_fc_1/kernel/Assign' id:16480 op device:{requested: '', assigned: ''} def:{{{node mrcnn_bbox_fc_1/kernel/Assign}} = AssignVariableOp[_has_manual_control_dependencies=true, dtype=DT_FLOAT, validate_shape=false](mrcnn_bbox_fc_1/kernel, mrcnn_bbox_fc_1/kernel/Initializer/stateless_random_uniform)}}' was changed by setting attribute after it was run by a session. This mutation will have no effect, and will trigger an error in the future. Either don't modify nodes after running them or create a new session.\n"
     ]
    },
    {
     "ename": "InvalidArgumentError",
     "evalue": "Graph execution error:\n\nDetected at node 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter' defined at (most recent call last):\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 86, in _run_code\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/events.py\", line 80, in _run\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n    File \"/var/folders/1k/lyyvyrld0x186lb492fcl7yr0000gn/T/ipykernel_58088/4009203555.py\", line 5, in <module>\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 1861, in __init__\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 2068, in build\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 993, in fpn_classifier_graph\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 814, in __call__\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 2182, in _maybe_build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/time_distributed.py\", line 137, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/base_wrapper.py\", line 57, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/core/dense.py\", line 154, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 459, in add_weight\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_utils.py\", line 137, in make_variable\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 647, in __call__\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 672, in _generate_init_val\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/backend.py\", line 2102, in random_uniform\nNode: 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter'\nDetected at node 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter' defined at (most recent call last):\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 86, in _run_code\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/events.py\", line 80, in _run\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n    File \"/var/folders/1k/lyyvyrld0x186lb492fcl7yr0000gn/T/ipykernel_58088/4009203555.py\", line 5, in <module>\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 1861, in __init__\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 2068, in build\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 993, in fpn_classifier_graph\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 814, in __call__\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 2182, in _maybe_build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/time_distributed.py\", line 137, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/base_wrapper.py\", line 57, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/core/dense.py\", line 154, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 459, in add_weight\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_utils.py\", line 137, in make_variable\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 647, in __call__\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 672, in _generate_init_val\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/backend.py\", line 2102, in random_uniform\nNode: 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter'\nCannot assign a device for operation mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter: Could not satisfy explicit device specification '' because the node {{colocation_node mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter}} was colocated with a group of nodes that required incompatible device '/job:localhost/replica:0/task:0/device:GPU:0'. All available devices [/job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:GPU:0]. \nColocation Debug Info:\nColocation group had the following types and supported devices: \nRoot Member(assigned_device_name_index_=-1 requested_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' resource_device_name_='' supported_device_types_=[CPU] possible_devices_=[]\nVarHandleOp: GPU CPU \nMul: GPU CPU \nAddV2: GPU CPU \nSub: GPU CPU \nAssignVariableOp: GPU CPU \nStatelessRandomGetKeyCounter: CPU \nStatelessRandomUniformV2: GPU CPU \nConst: GPU CPU \n\nColocation members, user-requested devices, and framework assigned devices, if any:\n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/shape (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/min (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/max (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter/seed (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter (StatelessRandomGetKeyCounter) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomUniformV2/alg (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomUniformV2 (StatelessRandomUniformV2) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/sub (Sub) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/mul (Mul) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform (AddV2) \n  mrcnn_class_logits_1/kernel (VarHandleOp)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0\n  mrcnn_class_logits_1/kernel/Assign (AssignVariableOp) \n\n\t [[{{node mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter}}]]\n\nOriginal stack trace for 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter':\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 86, in _run_code\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/events.py\", line 80, in _run\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n  File \"/var/folders/1k/lyyvyrld0x186lb492fcl7yr0000gn/T/ipykernel_58088/4009203555.py\", line 5, in <module>\n  File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 1861, in __init__\n  File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 2068, in build\n  File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 993, in fpn_classifier_graph\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 814, in __call__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 2182, in _maybe_build\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/time_distributed.py\", line 137, in build\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/base_wrapper.py\", line 57, in build\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/core/dense.py\", line 154, in build\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 459, in add_weight\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/trackable/base.py\", line 492, in _add_variable_with_custom_getter\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_utils.py\", line 137, in make_variable\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py\", line 150, in error_handler\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/variables.py\", line 195, in __call__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/variable_v1.py\", line 306, in _variable_call\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/variable_v1.py\", line 299, in <lambda>\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/ref_variable.py\", line 65, in default_variable_creator\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py\", line 150, in error_handler\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/variables.py\", line 198, in __call__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/resource_variable_ops.py\", line 1809, in __init__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/resource_variable_ops.py\", line 1993, in _init_from_args\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 647, in __call__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 672, in _generate_init_val\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/backend.py\", line 2102, in random_uniform\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py\", line 150, in error_handler\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/util/dispatch.py\", line 1260, in op_dispatch_handler\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/stateless_random_ops.py\", line 402, in stateless_random_uniform\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/random_ops_util.py\", line 149, in get_key_counter_alg\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/random_ops_util.py\", line 112, in _get_key_counter\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/gen_stateless_random_ops_v2.py\", line 195, in stateless_random_get_key_counter\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/framework/op_def_library.py\", line 796, in _apply_op_helper\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/framework/ops.py\", line 2657, in _create_op_internal\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/framework/ops.py\", line 1161, in from_node_def\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInvalidArgumentError\u001b[0m                      Traceback (most recent call last)",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/client/session.py:1402\u001b[0m, in \u001b[0;36mBaseSession._do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1401\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1402\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1403\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m errors\u001b[38;5;241m.\u001b[39mOpError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/client/session.py:1385\u001b[0m, in \u001b[0;36mBaseSession._do_run.<locals>._run_fn\u001b[0;34m(feed_dict, fetch_list, target_list, options, run_metadata)\u001b[0m\n\u001b[1;32m   1384\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_extend_graph()\n\u001b[0;32m-> 1385\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_tf_sessionrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfeed_dict\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfetch_list\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1386\u001b[0m \u001b[43m                                \u001b[49m\u001b[43mtarget_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_metadata\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/client/session.py:1478\u001b[0m, in \u001b[0;36mBaseSession._call_tf_sessionrun\u001b[0;34m(self, options, feed_dict, fetch_list, target_list, run_metadata)\u001b[0m\n\u001b[1;32m   1476\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_call_tf_sessionrun\u001b[39m(\u001b[38;5;28mself\u001b[39m, options, feed_dict, fetch_list, target_list,\n\u001b[1;32m   1477\u001b[0m                         run_metadata):\n\u001b[0;32m-> 1478\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtf_session\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mTF_SessionRun_wrapper\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_session\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moptions\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfeed_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1479\u001b[0m \u001b[43m                                          \u001b[49m\u001b[43mfetch_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtarget_list\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1480\u001b[0m \u001b[43m                                          \u001b[49m\u001b[43mrun_metadata\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mInvalidArgumentError\u001b[0m: Cannot assign a device for operation mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter: Could not satisfy explicit device specification '' because the node {{colocation_node mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter}} was colocated with a group of nodes that required incompatible device '/job:localhost/replica:0/task:0/device:GPU:0'. All available devices [/job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:GPU:0]. \nColocation Debug Info:\nColocation group had the following types and supported devices: \nRoot Member(assigned_device_name_index_=-1 requested_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' resource_device_name_='' supported_device_types_=[CPU] possible_devices_=[]\nVarHandleOp: GPU CPU \nMul: GPU CPU \nAddV2: GPU CPU \nSub: GPU CPU \nAssignVariableOp: GPU CPU \nStatelessRandomGetKeyCounter: CPU \nStatelessRandomUniformV2: GPU CPU \nConst: GPU CPU \n\nColocation members, user-requested devices, and framework assigned devices, if any:\n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/shape (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/min (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/max (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter/seed (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter (StatelessRandomGetKeyCounter) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomUniformV2/alg (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomUniformV2 (StatelessRandomUniformV2) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/sub (Sub) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/mul (Mul) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform (AddV2) \n  mrcnn_class_logits_1/kernel (VarHandleOp)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0\n  mrcnn_class_logits_1/kernel/Assign (AssignVariableOp) \n\n\t [[{{node mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter}}]]",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mInvalidArgumentError\u001b[0m                      Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 42\u001b[0m\n\u001b[1;32m     34\u001b[0m     \u001b[38;5;66;03m# Visualize the detected objects.\u001b[39;00m\n\u001b[1;32m     35\u001b[0m     mrcnn\u001b[38;5;241m.\u001b[39mvisualize\u001b[38;5;241m.\u001b[39mdisplay_instances(image\u001b[38;5;241m=\u001b[39mimg, \n\u001b[1;32m     36\u001b[0m                                     boxes\u001b[38;5;241m=\u001b[39mr[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrois\u001b[39m\u001b[38;5;124m'\u001b[39m], \n\u001b[1;32m     37\u001b[0m                                     masks\u001b[38;5;241m=\u001b[39mr[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmasks\u001b[39m\u001b[38;5;124m'\u001b[39m], \n\u001b[1;32m     38\u001b[0m                                     class_ids\u001b[38;5;241m=\u001b[39mr[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mclass_ids\u001b[39m\u001b[38;5;124m'\u001b[39m], \n\u001b[1;32m     39\u001b[0m                                     class_names\u001b[38;5;241m=\u001b[39mCLASS_NAMES, \n\u001b[1;32m     40\u001b[0m                                     scores\u001b[38;5;241m=\u001b[39mr[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mscores\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[0;32m---> 42\u001b[0m \u001b[43mdetect_and_color_splash\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimage_path\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdata2/train/3228694_Box_240506_145035_color.png\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[8], line 27\u001b[0m, in \u001b[0;36mdetect_and_color_splash\u001b[0;34m(model, image_path)\u001b[0m\n\u001b[1;32m     25\u001b[0m img \u001b[38;5;241m=\u001b[39m skimage\u001b[38;5;241m.\u001b[39mio\u001b[38;5;241m.\u001b[39mimread(image_path)\n\u001b[1;32m     26\u001b[0m \u001b[38;5;66;03m# Detect objects\u001b[39;00m\n\u001b[0;32m---> 27\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdetect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mimg\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m     28\u001b[0m \u001b[38;5;66;03m# Color splash\u001b[39;00m\n\u001b[1;32m     29\u001b[0m splash \u001b[38;5;241m=\u001b[39m color_splash(img, r[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmasks\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "File \u001b[0;32m~/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py:2548\u001b[0m, in \u001b[0;36mMaskRCNN.detect\u001b[0;34m(self, images, verbose)\u001b[0m\n\u001b[1;32m   2545\u001b[0m     log(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124manchors\u001b[39m\u001b[38;5;124m\"\u001b[39m, anchors)\n\u001b[1;32m   2546\u001b[0m \u001b[38;5;66;03m# Run object detection\u001b[39;00m\n\u001b[1;32m   2547\u001b[0m detections, _, _, mrcnn_mask, _, _, _ \u001b[38;5;241m=\u001b[39m\\\n\u001b[0;32m-> 2548\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkeras_model\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpredict\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mmolded_images\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimage_metas\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43manchors\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2549\u001b[0m \u001b[38;5;66;03m# Process detections\u001b[39;00m\n\u001b[1;32m   2550\u001b[0m results \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/training_v1.py:1059\u001b[0m, in \u001b[0;36mModel.predict\u001b[0;34m(self, x, batch_size, verbose, steps, callbacks, max_queue_size, workers, use_multiprocessing)\u001b[0m\n\u001b[1;32m   1056\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_call_args(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpredict\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1058\u001b[0m func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_select_training_loop(x)\n\u001b[0;32m-> 1059\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpredict\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1060\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1061\u001b[0m \u001b[43m    \u001b[49m\u001b[43mx\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1062\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbatch_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbatch_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1063\u001b[0m \u001b[43m    \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1064\u001b[0m \u001b[43m    \u001b[49m\u001b[43msteps\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msteps\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1065\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1066\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmax_queue_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmax_queue_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1067\u001b[0m \u001b[43m    \u001b[49m\u001b[43mworkers\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mworkers\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1068\u001b[0m \u001b[43m    \u001b[49m\u001b[43muse_multiprocessing\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_multiprocessing\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1069\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/training_arrays_v1.py:801\u001b[0m, in \u001b[0;36mArrayLikeTrainingLoop.predict\u001b[0;34m(self, model, x, batch_size, verbose, steps, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    797\u001b[0m batch_size \u001b[38;5;241m=\u001b[39m model\u001b[38;5;241m.\u001b[39m_validate_or_infer_batch_size(batch_size, steps, x)\n\u001b[1;32m    798\u001b[0m x, _, _ \u001b[38;5;241m=\u001b[39m model\u001b[38;5;241m.\u001b[39m_standardize_user_data(\n\u001b[1;32m    799\u001b[0m     x, check_steps\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, steps_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msteps\u001b[39m\u001b[38;5;124m\"\u001b[39m, steps\u001b[38;5;241m=\u001b[39msteps\n\u001b[1;32m    800\u001b[0m )\n\u001b[0;32m--> 801\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mpredict_loop\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    802\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    803\u001b[0m \u001b[43m    \u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    804\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbatch_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbatch_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    805\u001b[0m \u001b[43m    \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    806\u001b[0m \u001b[43m    \u001b[49m\u001b[43msteps\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msteps\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    807\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    808\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/training_arrays_v1.py:421\u001b[0m, in \u001b[0;36mmodel_iteration\u001b[0;34m(model, inputs, targets, sample_weights, batch_size, epochs, verbose, callbacks, val_inputs, val_targets, val_sample_weights, shuffle, initial_epoch, steps_per_epoch, validation_steps, validation_freq, mode, validation_in_fit, prepared_feed_values_from_dataset, steps_name, **kwargs)\u001b[0m\n\u001b[1;32m    416\u001b[0m callbacks\u001b[38;5;241m.\u001b[39m_call_batch_hook(\n\u001b[1;32m    417\u001b[0m     mode, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbegin\u001b[39m\u001b[38;5;124m\"\u001b[39m, batch_index, batch_logs\n\u001b[1;32m    418\u001b[0m )\n\u001b[1;32m    420\u001b[0m \u001b[38;5;66;03m# Get outputs.\u001b[39;00m\n\u001b[0;32m--> 421\u001b[0m batch_outs \u001b[38;5;241m=\u001b[39m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[43mins_batch\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    422\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(batch_outs, \u001b[38;5;28mlist\u001b[39m):\n\u001b[1;32m    423\u001b[0m     batch_outs \u001b[38;5;241m=\u001b[39m [batch_outs]\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/backend.py:4567\u001b[0m, in \u001b[0;36mGraphExecutionFunction.__call__\u001b[0;34m(self, inputs)\u001b[0m\n\u001b[1;32m   4561\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__call__\u001b[39m(\u001b[38;5;28mself\u001b[39m, inputs):\n\u001b[1;32m   4562\u001b[0m     inputs \u001b[38;5;241m=\u001b[39m tf\u001b[38;5;241m.\u001b[39mnest\u001b[38;5;241m.\u001b[39mflatten(\n\u001b[1;32m   4563\u001b[0m         tf_utils\u001b[38;5;241m.\u001b[39mconvert_variables_to_tensors(inputs),\n\u001b[1;32m   4564\u001b[0m         expand_composites\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m   4565\u001b[0m     )\n\u001b[0;32m-> 4567\u001b[0m     session \u001b[38;5;241m=\u001b[39m \u001b[43mget_session\u001b[49m\u001b[43m(\u001b[49m\u001b[43minputs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4568\u001b[0m     feed_arrays \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m   4569\u001b[0m     array_vals \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/backend.py:786\u001b[0m, in \u001b[0;36mget_session\u001b[0;34m(op_input_list)\u001b[0m\n\u001b[1;32m    784\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m _MANUAL_VAR_INIT:\n\u001b[1;32m    785\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m session\u001b[38;5;241m.\u001b[39mgraph\u001b[38;5;241m.\u001b[39mas_default():\n\u001b[0;32m--> 786\u001b[0m         \u001b[43m_initialize_variables\u001b[49m\u001b[43m(\u001b[49m\u001b[43msession\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    787\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m session\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/backend.py:1267\u001b[0m, in \u001b[0;36m_initialize_variables\u001b[0;34m(session)\u001b[0m\n\u001b[1;32m   1265\u001b[0m     v\u001b[38;5;241m.\u001b[39m_keras_initialized \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m   1266\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m uninitialized_vars:\n\u001b[0;32m-> 1267\u001b[0m     \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompat\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mv1\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvariables_initializer\u001b[49m\u001b[43m(\u001b[49m\u001b[43muninitialized_vars\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/client/session.py:972\u001b[0m, in \u001b[0;36mBaseSession.run\u001b[0;34m(self, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m    969\u001b[0m run_metadata_ptr \u001b[38;5;241m=\u001b[39m tf_session\u001b[38;5;241m.\u001b[39mTF_NewBuffer() \u001b[38;5;28;01mif\u001b[39;00m run_metadata \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    971\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 972\u001b[0m   result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfetches\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfeed_dict\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moptions_ptr\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    973\u001b[0m \u001b[43m                     \u001b[49m\u001b[43mrun_metadata_ptr\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    974\u001b[0m   \u001b[38;5;28;01mif\u001b[39;00m run_metadata:\n\u001b[1;32m    975\u001b[0m     proto_data \u001b[38;5;241m=\u001b[39m tf_session\u001b[38;5;241m.\u001b[39mTF_GetBuffer(run_metadata_ptr)\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/client/session.py:1215\u001b[0m, in \u001b[0;36mBaseSession._run\u001b[0;34m(self, handle, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1212\u001b[0m \u001b[38;5;66;03m# We only want to really perform the run if fetches or targets are provided,\u001b[39;00m\n\u001b[1;32m   1213\u001b[0m \u001b[38;5;66;03m# or if the call is a partial run that specifies feeds.\u001b[39;00m\n\u001b[1;32m   1214\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m final_fetches \u001b[38;5;129;01mor\u001b[39;00m final_targets \u001b[38;5;129;01mor\u001b[39;00m (handle \u001b[38;5;129;01mand\u001b[39;00m feed_dict_tensor):\n\u001b[0;32m-> 1215\u001b[0m   results \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_do_run\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfinal_targets\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfinal_fetches\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1216\u001b[0m \u001b[43m                         \u001b[49m\u001b[43mfeed_dict_tensor\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moptions\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_metadata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1217\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1218\u001b[0m   results \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/client/session.py:1395\u001b[0m, in \u001b[0;36mBaseSession._do_run\u001b[0;34m(self, handle, target_list, fetch_list, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1392\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_call_tf_sessionprun(handle, feed_dict, fetch_list)\n\u001b[1;32m   1394\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m handle \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1395\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_do_call\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_run_fn\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfeeds\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfetches\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtargets\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moptions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1396\u001b[0m \u001b[43m                       \u001b[49m\u001b[43mrun_metadata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1397\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1398\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_do_call(_prun_fn, handle, feeds, fetches)\n",
      "File \u001b[0;32m~/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/client/session.py:1421\u001b[0m, in \u001b[0;36mBaseSession._do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1416\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124monly supports NHWC tensor format\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m message:\n\u001b[1;32m   1417\u001b[0m   message \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m (\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mA possible workaround: Try disabling Grappler optimizer\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m   1418\u001b[0m               \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mby modifying the config for creating the session eg.\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m   1419\u001b[0m               \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124msession_config.graph_options.rewrite_options.\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m   1420\u001b[0m               \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdisable_meta_optimizer = True\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m-> 1421\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(e)(node_def, op, message)\n",
      "\u001b[0;31mInvalidArgumentError\u001b[0m: Graph execution error:\n\nDetected at node 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter' defined at (most recent call last):\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 86, in _run_code\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/events.py\", line 80, in _run\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n    File \"/var/folders/1k/lyyvyrld0x186lb492fcl7yr0000gn/T/ipykernel_58088/4009203555.py\", line 5, in <module>\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 1861, in __init__\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 2068, in build\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 993, in fpn_classifier_graph\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 814, in __call__\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 2182, in _maybe_build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/time_distributed.py\", line 137, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/base_wrapper.py\", line 57, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/core/dense.py\", line 154, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 459, in add_weight\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_utils.py\", line 137, in make_variable\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 647, in __call__\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 672, in _generate_init_val\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/backend.py\", line 2102, in random_uniform\nNode: 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter'\nDetected at node 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter' defined at (most recent call last):\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 86, in _run_code\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/events.py\", line 80, in _run\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n    File \"/var/folders/1k/lyyvyrld0x186lb492fcl7yr0000gn/T/ipykernel_58088/4009203555.py\", line 5, in <module>\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 1861, in __init__\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 2068, in build\n    File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 993, in fpn_classifier_graph\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 814, in __call__\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 2182, in _maybe_build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/time_distributed.py\", line 137, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/base_wrapper.py\", line 57, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/core/dense.py\", line 154, in build\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 459, in add_weight\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_utils.py\", line 137, in make_variable\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 647, in __call__\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 672, in _generate_init_val\n    File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/backend.py\", line 2102, in random_uniform\nNode: 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter'\nCannot assign a device for operation mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter: Could not satisfy explicit device specification '' because the node {{colocation_node mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter}} was colocated with a group of nodes that required incompatible device '/job:localhost/replica:0/task:0/device:GPU:0'. All available devices [/job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:GPU:0]. \nColocation Debug Info:\nColocation group had the following types and supported devices: \nRoot Member(assigned_device_name_index_=-1 requested_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' assigned_device_name_='/job:localhost/replica:0/task:0/device:GPU:0' resource_device_name_='' supported_device_types_=[CPU] possible_devices_=[]\nVarHandleOp: GPU CPU \nMul: GPU CPU \nAddV2: GPU CPU \nSub: GPU CPU \nAssignVariableOp: GPU CPU \nStatelessRandomGetKeyCounter: CPU \nStatelessRandomUniformV2: GPU CPU \nConst: GPU CPU \n\nColocation members, user-requested devices, and framework assigned devices, if any:\n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/shape (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/min (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/max (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter/seed (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter (StatelessRandomGetKeyCounter) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomUniformV2/alg (Const) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomUniformV2 (StatelessRandomUniformV2) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/sub (Sub) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/mul (Mul) \n  mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform (AddV2) \n  mrcnn_class_logits_1/kernel (VarHandleOp)  framework assigned device=/job:localhost/replica:0/task:0/device:GPU:0\n  mrcnn_class_logits_1/kernel/Assign (AssignVariableOp) \n\n\t [[{{node mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter}}]]\n\nOriginal stack trace for 'mrcnn_class_logits_1/kernel/Initializer/stateless_random_uniform/StatelessRandomGetKeyCounter':\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/runpy.py\", line 86, in _run_code\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/asyncio/events.py\", line 80, in _run\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n  File \"/var/folders/1k/lyyvyrld0x186lb492fcl7yr0000gn/T/ipykernel_58088/4009203555.py\", line 5, in <module>\n  File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 1861, in __init__\n  File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 2068, in build\n  File \"/Users/basti/Documents/maqum/Mask-RCNN_TF2.14.0/mrcnn/model.py\", line 993, in fpn_classifier_graph\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 814, in __call__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 2182, in _maybe_build\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/time_distributed.py\", line 137, in build\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/rnn/base_wrapper.py\", line 57, in build\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/layers/core/dense.py\", line 154, in build\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_v1.py\", line 459, in add_weight\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/trackable/base.py\", line 492, in _add_variable_with_custom_getter\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/engine/base_layer_utils.py\", line 137, in make_variable\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py\", line 150, in error_handler\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/variables.py\", line 195, in __call__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/variable_v1.py\", line 306, in _variable_call\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/variable_v1.py\", line 299, in <lambda>\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/ref_variable.py\", line 65, in default_variable_creator\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py\", line 150, in error_handler\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/variables.py\", line 198, in __call__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/resource_variable_ops.py\", line 1809, in __init__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/resource_variable_ops.py\", line 1993, in _init_from_args\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 647, in __call__\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/initializers/initializers.py\", line 672, in _generate_init_val\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/keras/src/backend.py\", line 2102, in random_uniform\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py\", line 150, in error_handler\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/util/dispatch.py\", line 1260, in op_dispatch_handler\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/stateless_random_ops.py\", line 402, in stateless_random_uniform\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/random_ops_util.py\", line 149, in get_key_counter_alg\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/random_ops_util.py\", line 112, in _get_key_counter\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/ops/gen_stateless_random_ops_v2.py\", line 195, in stateless_random_get_key_counter\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/framework/op_def_library.py\", line 796, in _apply_op_helper\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/framework/ops.py\", line 2657, in _create_op_internal\n  File \"/Users/basti/.pyenv/versions/miniforge3-23.11.0-0/lib/python3.10/site-packages/tensorflow/python/framework/ops.py\", line 1161, in from_node_def\n"
     ]
    }
   ],
   "source": [
    "#Show detected objects in color and all others in B&W    \n",
    "def color_splash(img, mask):\n",
    "    \"\"\"Apply color splash effect.\n",
    "    image: RGB image [height, width, 3]\n",
    "    mask: instance segmentation mask [height, width, instance count]\n",
    "    Returns result image.\n",
    "    \"\"\"\n",
    "    # Make a grayscale copy of the image. The grayscale copy still\n",
    "    # has 3 RGB channels, though.\n",
    "    gray = skimage.color.gray2rgb(skimage.color.rgb2gray(img)) * 255\n",
    "    # Copy color pixels from the original color image where mask is set\n",
    "    if mask.shape[-1] > 0:\n",
    "        # We're treating all instances as one, so collapse the mask into one layer\n",
    "        mask = (np.sum(mask, -1, keepdims=True) >= 1)\n",
    "        splash = np.where(mask, img, gray).astype(np.uint8)\n",
    "    else:\n",
    "        splash = gray.astype(np.uint8)\n",
    "    return splash\n",
    "\n",
    "def detect_and_color_splash(model, image_path=None):\n",
    "    assert image_path\n",
    "    # Run model detection and generate the color splash effect\n",
    "    #print(\"Running on {}\".format(img))\n",
    "    # Read image\n",
    "    img = skimage.io.imread(image_path)\n",
    "    # Detect objects\n",
    "    r = model.detect([img], verbose=1)[0]\n",
    "    # Color splash\n",
    "    splash = color_splash(img, r['masks'])\n",
    "    # Save output\n",
    "    file_name = \"splash_{:%Y%m%dT%H%M%S}.png\".format(datetime.datetime.now())\n",
    "    skimage.io.imsave(file_name, splash)\n",
    "\n",
    "    # Visualize the detected objects.\n",
    "    mrcnn.visualize.display_instances(image=img, \n",
    "                                    boxes=r['rois'], \n",
    "                                    masks=r['masks'], \n",
    "                                    class_ids=r['class_ids'], \n",
    "                                    class_names=CLASS_NAMES, \n",
    "                                    scores=r['scores'])\n",
    "\n",
    "detect_and_color_splash(model, image_path=\"data2/train/3228694_Box_240506_145035_color.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
